### PR TITLE
Add Prometheus monitoring endpoint

### PR DESCRIPTION
**Pull request by dgl**
# Add Prometheus monitoring endpoint

Adds an optional HTTP server, with a metrics endpoint available on `/metrics`. This allows for Prometheus compatible metrics to be collected; currently it adds no metrics specific to seccompagent, but those could be added later.

This adds a dependency on github.com/prometheus/client_golang, I've updated vendoring as a result (in a different commit, ideally don't squash merge this to keep the history clearer).

## How to use

Run seccompagent with the monitor listen option set to a port.

## Testing done

```shell
$ ./seccompagent -monitor-listen :9929
$ curl localhost:9929/metrics
[returns metrics]
```

### CODE REVIEW

File: cmd/seccompagent/seccompagent.go [hunk 1]
```diff
@@ -41,12 +44,14 @@ var (
...
 	resolverParam string
 	logflags      string
+	monitorListen string
 )
 
 func init() {
 	flag.StringVar(&socketFile, "socketfile", "/run/seccomp-agent.socket", "Socket file")
 	flag.StringVar(&resolverParam, "resolver", "", "Container resolver to use [none, demo-basic, kubernetes]")
 	flag.StringVar(&logflags, "log", "info", "log level [trace,debug,info,warn,error,fatal,color,nocolor,json]")
+	flag.StringVar(&monitorListen, "monitor-listen", "", "[host]:port to listen on for monitoring, empty means no monitoring port")
```
Comment (alban) [inline cmd/seccompagent/seccompagent.go:54]: I see that many projects use `--metrics-bind-address`.

For example, [kube-proxy](https://kubernetes.io/docs/reference/command-line-tools-reference/kube-proxy/). Could you use the same flag name?

```diff
-	flag.StringVar(&monitorListen, "monitor-listen", "", "[host]:port to listen on for monitoring, empty means no monitoring port")
+	flag.StringVar(&metricsBindAddr, "metrics-bind-address", "", "The IP address with port for the metrics server to serve on (set to '0.0.0.0:10249' for all IPv4 interfaces and '[::]:10249' for all IPv6 interfaces). Set empty to disable.")
```

**Review by alban — COMMENTED** [review]
Do you have ideas of metrics to add?

**Comment by dgl** [comment]
Sorry for the delay; I've now added a middleware to have include metrics on the syscalls made and a small histogram.

Note this is useful even without the middleware, just being able to monitor the process is running via metrics and the Go level metrics can be useful.

**Comment by stackedsax** [comment]
Hey there @alban, any other feedback for @dgl on this?

File: pkg/handlers/prometheus/promtheus.go [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+			syscallName, err := req.Data.Syscall.GetName()
+			if err != nil {
+				log.WithFields(log.Fields{
+					"fd":  fd,
+					"req": req,
+					"err": err,
+				}).Error("Error in decoding syscall")
+			}
+
+			r := h(fd, req)
```
Comment (alban) [inline pkg/handlers/prometheus/promtheus.go:59]: I think this is not measuring the duration of the syscall by the target process, but how long the seccomp agent takes to emulate the syscall (if such a handler is used for emulating the syscall).

I am curious about your use case and what syscall you emulate.

**Review by alban — APPROVED** [review]
LGTM, thanks.

I am pondering removing the vendor directory from git. But that's unrelated to this PR.

**Pull request merged**